File: android/src/main/kotlin/com/situm/flutter/ar/situm_ar/ARMethodCallHandler.kt
package com.situm.flutter.ar.situm_ar

import android.util.Log
import io.flutter.plugin.common.MethodChannel

/**
 * MethodCall wrapper/adapter. Use this class to handle Dart messages.
 */
class ARMethodCallHandler(
    private val controller: ARController
) {
    companion object {
        const val TAG = "Situm> AR>"
        const val DONE = "DONE"
    }

    fun handle(method: String?, arguments: Map<String, Any>, result: MethodChannel.Result) {
        when (method) {
            "load" -> handleLoad(arguments, result)
            "pause" -> handlePause(arguments, result)
            "resume" -> handleResume(arguments, result)
            "unload" -> handleUnload(arguments, result)
            "worldRedraw" -> handleRedraw(arguments, result)
            "updateArrowTarget" -> handleUpdateArrowTarget(arguments, result)
            "getDebugInfo" -> handleGetDebugInfo(arguments, result)
            else -> result.notImplemented()
        }
    }

    private fun handleLoad(arguments: Map<String, Any>, result: MethodChannel.Result) {
        val buildingIdentifier = arguments["buildingIdentifier"] as String
        controller.load(buildingIdentifier)
        result.success(DONE)
        Log.d(TAG, "### AR has been LOADED and should be visible ###")
    }

    private fun handleUnload(arguments: Map<String, Any>, result: MethodChannel.Result) {
        controller.unload()
        result.success(DONE)
        Log.d(TAG, "### AR has been UNLOADED ###")
    }

    private fun handleResume(arguments: Map<String, Any>, result: MethodChannel.Result) {
        controller.resume()
        result.success(DONE)
        Log.d(TAG, "### AR has been RESUMED ###")
    }

    private fun handlePause(arguments: Map<String, Any>, result: MethodChannel.Result) {
        controller.pause()
        result.success(DONE)
        Log.d(TAG, "### AR has been PAUSED (camera should not be active) ###")
    }

    private fun handleRedraw(arguments: Map<String, Any>, result: MethodChannel.Result) {
        controller.worldRedraw()
        result.success(DONE)
    }

    private fun handleUpdateArrowTarget(arguments: Map<String, Any>, result: MethodChannel.Result) {
        controller.updateArrowTarget()
        result.success(DONE)
    }

    private fun handleGetDebugInfo(arguments: Map<String, Any>, result: MethodChannel.Result){
        result.success(controller.getDebugInfo())
    }


}

File: ios/Classes/ARMethodCallHandler.swift
import Foundation
import Flutter

@available(iOS 14.0, *)
class ARMethodCallHandler {
    
    private let controller: ARController
    
    init(arController: ARController) {
        self.controller = arController
    }
    
    // Constantes
    private let TAG = "Situm> AR>"
    private let DONE = "DONE"
    
    // Método para xestionar chamadas de métodos desde Dart
    func handle(method: String, arguments: [String: Any], result: @escaping FlutterResult) {        
        switch method {
        case "load":
            handleLoad(arguments: arguments, result: result)
        case "pause":
            handlePause(arguments: arguments, result: result)
        case "resume":
            handleResume(arguments: arguments, result: result)
        case "unload":
            handleUnload(arguments: arguments, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func handleLoad(arguments: [String: Any], result: @escaping FlutterResult) {
        let buildingIdentifer = arguments["buildingIdentifier"] as! String
        controller.load(buildingIdentifier: buildingIdentifer)
        result(DONE)
        print("\(TAG) ### AR has been LOADED and should be visible ###")
    }
    
    private func handleUnload(arguments: [String: Any], result: @escaping FlutterResult) {
        controller.unload()
        result(DONE)
        print("\(TAG) ### AR has been UNLOADED ###")
    }
    
    private func handleResume(arguments: [String: Any], result: @escaping FlutterResult) {
        controller.resume()
        result(DONE)
        print("\(TAG) ### AR has been RESUMED ###")
    }
    
    private func handlePause(arguments: [String: Any], result: @escaping FlutterResult) {
        controller.pause()
        result(DONE)
        print("\(TAG) ### AR has been PAUSED (camera should not be active) ###")
    }
}
